Xarray GPU optimization (#771)
xarray-contrib/xarray.dev: src/posts/gpu-pipeline/index.md

---

> title: 'Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)'

Reviewer: ```suggestion
title: 'GPU-Native Earth Science AI/ML Workflows Xarray, Zarr, DALI, and nvcomp'
```

date: '2025-05-01'

authors:
  - name: Negin Sobhani
    github: negin513
  - name: Wei Ji Leong
    github: weiji14
  - name: Max Jones
    github: maxjones
  - name: Akshay Subranian
    github: akshaysubr
  - name: Thomas Augspurger
    github: tomaugspurger
  - name: Katelyn Fitzgerald
    github: kafitzgerald

> summary: 'How to accelerate AI/ML workflows in Earth Sciences with GPU-native Xarray and Zarr.'

Reviewer: Can we make this more direct? "X% speedup" or "XMBps throughput"?

Author: Added ~17X all in all improvement!💪💪💪

---

# Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)

> 

Reviewer: Let's add a TL; DR

## Introduction

> In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to transfer massive datasets from disk to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.

Reviewer 2: ```suggestion
In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to preprocess and transfer massive datasets from storage to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.
```

> To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as Zarr, kvikIO, and DALI.

Reviewer 2: ```suggestion
To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs.io/), [KvikIO](https://docs.rapids.ai/api/kvikio/stable/), and [DALI](https://developer.nvidia.com/dali).
```

> To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in a GPU hackathon to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs.io/), [KvikIO](https://docs.rapids.ai/api/kvikio/stable/), and [DALI](https://developer.nvidia.com/dali).

Reviewer: Some institution logos here could be nice visually

Author: Good Idea! Added!

Author: BTW, you meant the logo of NCAR, devseed, etc? That is added now at the end of the post, but I am open to bring them higher. 

I can also include the logos of packages( i.e. cupy, etc.) but not sure if that was what you meant.

> In this post, we share our hackathon experience, the integration strategies we explored, and the performance gains we achieved to highlight how modern tools can transform data-intensive workflows.

Reviewer: Let's call out the hackathon and its funders by name (useful in this _climate_ hehe)

Reviewer: I see this at the end, but it would be nice to call it out up top

Author: Done. Thanks for the suggestion. 🥇

Author: 100% agreed. Linked them at the beginning of the blog post and would be happy to highlight them more too. :-)

## Problem

Machine learning pipelines typically involve:

- Reading data from disk or object storage.
- Transforming / preprocessing data (often CPU-bound).
- Feeding the data into GPUs for training or inference.

> Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large dataset.

Reviewer 2: ```suggestion
Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets.
```

> Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets.

Reviewer: ```suggestion
Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets (profiles discussed below).
```

In this hackathon, we tried looking at different ways of reducing this bottleneck.

### Data & Code Overview

> For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) training on ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.

Reviewer 2: ```suggestion
For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) model on the ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.
```

> For this hackathon, we developed a benchmark of training a U-NET (with ResNet backend) model on the ERA-5 Dataset to predict next time steps. The U-Net model is implemented in PyTorch and the training pipeline is built using PyTorch DataLoader. The model can be trained on a single GPU or multiple GPUs using Distributed Data Parallel (DDP) for parallelization.

Reviewer: Let's link to the Repo here before referring to files

> -- TODO : Add an example image.

Reviewer: flagging this TODO

The basic data loader is implemented in `zarr_ML_optimization/train_unet.py` and the model is defined in `zarr_ML_optimization/model.py`. The training pipeline is designed to be flexible and can be easily adapted to different datasets and models.

More details on the model and training pipeline can be found in the [README](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/zarr_ML_optimization/README.md) file in the `zarr_ML_optimization` folder.

### Performance Bottlenecks

First, we needed to identify the performance bottlenecks in our pipeline. We used NVIDIA's [Nsight Systems](https://developer.nvidia.com/nsight-systems) to profile our code and identify the areas that needed optimization.

Here are some screenshots of the profiling results:

> - ![image](https://hackmd.io/_uploads/H1N1TTtAyl.png)

Reviewer 3: @negin513, could you provide push permissions to your fork? I can help copy the images from hackmd to here.

> - ![image](https://hackmd.io/_uploads/H1N1TTtAyl.png)

Author: @weiji14 done. ✅ 

Thanks @dcherian , @TomAugspurger , and @weiji14 for your comments and for reviewing this post.

This was also confirmed by a few other flags we added in our script to measure the time spent on data loading and model training. The results are shown below:

![baseline plot](baseline.png)

In the plot above, we show the throughput of the data loading and training steps in our pipeline. The three bars represent:

- Real Data: Baseline throughput of the end-to-end pipeline using real data.
- No Training (i.e. data loading throughput): Throughput of the data loading without any training (to measure the time spent on data loading vs. training).
- Synthetic Data (i.e. Training throughput): Throughput of the data loading using synthetic data (to remove the data loading bottleneck).

The results show that the data loading step is the main bottleneck in our pipeline, with **much** lower throughput compared to the training step.

## Hackathon: Putting this altogether

Our initial profiling showed that data loading is a major bottleneck in this workflow.

During the hackathon, we tested the following strategies to improve the data loading performance:

1. Optimized Chunking & Compression

>    - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.

Reviewer: I would either delete this summary or shorten and add a summary to the remaining items.

Some links to nvcomp and DALI would be great too

### Step 1: Optimized chunking :card_file_box:

The ERA-5 dataset we were using had a sub-optimal chunking scheme of `{'time': 10, 'channel': C, 'height': H, 'width': W}`, which meant that a minimum of 10 timesteps of data was being read even if we only needed 2 consecutive timesteps at a time.
We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr v3.
The full script is available [here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/rechunk/era5_rechunking.ipynb), with the main code looking like so:

```python
import xarray as xr

ds: xr.Dataset = xr.open_mfdataset("ERA5.zarr")
# Rechunk the data
ds = ds.chunk({"time": 1, "level": 1, "latitude": 640, "longitude": 1280})
# Save to Zarr v3
ds.to_zarr("rechunked_ERA5.zarr", zarr_version=3)
```

For more optimal performance, consider:

1. Storing the data without compression (if not transferring over a network), as decompressing data can slow down read speeds. But see also GPU decompression with nvCOMP below. :wink:

> 2. Concatenating several data variables together **if** a single chunk size is too small (<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, <10MB (??TODO verify) for optimal reads.

Reviewer 3: Need to escape the "<" character here to fix build error "Unexpected character `1` (U+0031) before name, expected a character that can start a name, such as a letter, `$`, or `_`".

```suggestion
2. Concatenating several data variables together **if** a single chunk size is too small (\<1MB), at the expense of reducing readability of the Zarr store. Having too many small chunks can be detrimental to read speeds. A compressed chunk should be >1MB, \<10MB (??TODO verify) for optimal reads.
```

Also, reminder that we need to verify if these chunk sizes (1MB, 10MB) are correct, my memory is a bit foggy now :face_in_clouds:

Author: Thanks! I spent some time trying to figure what went wrong in my local deployment and then I say your comment. 😉

Author: I changed this to be between 1MB and 100MB. 5-100 MB per chunk is along the line of what we recommend for lustre. Our chunk sizes were around 6.25 MiB.

```

Note that using `engine="zarr"` like above would still result in data being loaded into CPU memory before it goes to GPU memory.
If you prefer to bypass CPU memory, and have GPU Direct Storage (GDS) enabled, you can use the `kvikio` driver like so:

```python
import kvikio.zarr

with zarr.config.enable_gpu():
    store = kvikio.zarr.GDSStore(root="/tmp/air-temp.zarr")
    ds = xr.open_dataset(filename_or_obj=store, engine="zarr")
    assert isinstance(ds.air.data, cp.ndarray)
```

This will read the data directly from the Zarr store to GPU memory, bypassing CPU memory altogether. This is especially useful for large datasets, as it reduces the amount of data that needs to be transferred between CPU and GPU memory.

[ TODO: add a figure showing this -- technically decompression is still done on CPU. ]

(TODO ongoing work) Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:

```python
import cupy_xarray

ds = xr.open_dataset(filename_or_obj="/tmp/air-temp.zarr", engine="kvikio")
assert isinstance(ds.air.data, cp.ndarray)

> ```

Reviewer: This could go in a future work section at the end

Reviewer 3: Yeah, I'm not sure if this API is feasible or even desirable (have tried to implement this in https://github.com/xarray-contrib/cupy-xarray/pull/70, but no luck yet patching the buffer protocol). So ok to move this towards the end.

![image](https://hackmd.io/_uploads/H1kJkk9Akx.png)

Figure above shows benchmark comparing CPU vs GPU-based decompression, with or without GDS enabled.

Keep an eye on this space!

### Step 4: Overlapping CPU and GPU compute with NVIDIA DALI :twisted_rightwards_arrows:

Ideally, we want to minimize pauses where the device (GPU) is waiting for the host (CPU) or vice versa. This is one of the reasons we went with [NVIDIA DALI](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/index.html) that enables overlapping CPU and GPU computation.

Our full codebase is available at https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus for reference.
To start, there is a [zarr_DALI](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_DALI) folder with short, contained examples of a DALI pipeline loading from Zarr.

Next, look at the [zarr_ML_optimization](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_ML_optimization) folder that contains an end-to-end example on how this DALI pipeline can be integrated into a Pytorch Dataloader and full training workflow.

![image](https://hackmd.io/_uploads/H1YVp6tR1l.png)

> (TODO insert better nsight profiling figure than above showing overlapping CPU and GPU compute)

Reviewer: that would be really nice!

## Going Forward

This work is still ongoing, and we are continuing to explore ways to optimize data loading and processing for large-scale geospatial AI/ML workflows. We started this work during a 3-day hackathon, and we are excited to continue this work in the future. During the hackathon, we were able to make significant progress in optimizing data loading and processing for large-scale geospatial AI/ML workflows.

We are continuing to explore the following areas:

- GPU Direct Storage (GDS) for optimal performance
- NVIDIA DALI
- Work out how to use GDS when reading from cloud object store instead of on-prem disk.

> - etc

Reviewer 3: Want to shout out that reading/writing Zarr shards with GPU buffers (thanks @maxrjones and @TomAugspurger!) at https://github.com/zarr-developers/zarr-python/pull/2978 was just merged, and could go in here or somewhere above, depending on when this blog post gets published.

Author: Next, we should test sharding on Lustre with and w/o GPU buffers. 🙈

## Lessons Learned

- Chunking matters! It really does.
- Consider using GPU Direct Storage (GDS) for optimal performance, but be aware of the setup and configuration required.
- GPU Direct Storage (GDS) can be an improvement for data-intensive workflows, but requires some setup and configuration.
- NVIDIA DALI is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.

> - GPU-based decompression is a promising area for future work, but requires further development and testing.

Reviewer: Icechunk!

---
title: 'Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)'
date: '2025-05-01'

authors:
  - name: Negin Sobhani
    github: negin513
  - name: Wei Ji Leong
    github: weiji14

>   - name: Max Jones
>     github: maxjones

Reviewer: ```suggestion
  - name: Max Jones
    github: maxrjones
```

---
title: 'Accelerating AI/ML Workflows in Earth Sciences with GPU-Native Xarray and Zarr (and more!)'
date: '2025-05-01'

authors:
  - name: Negin Sobhani
    github: negin513
  - name: Wei Ji Leong
    github: weiji14
  - name: Max Jones
    github: maxrjones

>   - name: Akshay Subranian

Reviewer 3: ```suggestion
  - name: Akshay Subramaniam
```

## Introduction

In large-scale geospatial AI and machine learning workflows, data loading is often the main bottleneck. Traditional pipelines rely on CPUs to preprocess and transfer massive datasets from storage to GPU memory, consuming resources and limiting scalability and effective use of GPU resources.

To tackle this issue, a team from the [National Center for Atmospheric Research (NSF-NCAR)](https://ncar.ucar.edu) and [Development Seed](https://developmentseed.org) with mentors from [NVIDIA](https://www.nvidia.com) participated in the [OpenHackathon](https://www.openhackathons.org/s/) to demonstrate how AI/ML workflows in Earth system sciences can benefit from GPU-native workflows using tools such as [Zarr](https://zarr.readthedocs.io/), [KvikIO](https://docs.rapids.ai/api/kvikio/stable/), and [DALI](https://developer.nvidia.com/dali).

In this post, we share our hackathon experience, the integration strategies we explored, and the performance gains we achieved to highlight how modern tools can transform data-intensive workflows.

## Problem

ML pipelines for large scientific datasets typically include steps:

> - Reading raw data from disk or object storage (often CPU-bound)

Reviewer 3: ```suggestion
- Reading raw data from disk or object storage (often IO-bound)
```

Author: Good catch!

- Transforming / preprocessing data (often CPU-bound)
- Model Training/Inference (often GPU-bound)

> Although GPU compute is incredibly fast, the CPU can become a bottleneck when dealing with large datasets. In an ideal scenario, we want to saturate the GPU with data as quickly as possible to minimize idle time on both the CPU and GPU.

Reviewer 3: ```suggestion
Although GPU compute is incredibly fast, IO and CPU bottlenecks can be a pain when dealing with large datasets. In an ideal scenario, we want to saturate the GPU with data as quickly as possible to minimize idle time on both the CPU and GPU.
```

> In this hackathon, we explored several strategies to reduce the data loading bottleneck and build a GPU-native pipeline to maximize GPU utilization.

Author: Slightly reworded to address. ;-)

with zarr.config.enable_gpu():
    ds = xr.open_dataset("/tmp/air-temp.zarr", engine="zarr", consolidated=False)
    assert isinstance(ds.air.data, cp.ndarray)
```

⚠️ Note that using `engine="zarr"` like above would still result in data being loaded into CPU memory before being transferred to GPU memory.

II. **Option 2: Direct-to-GPU via KvikIO (Zarr -> GPU)**
If your system supports [GPU Direct Storage (GDS)](https://developer.nvidia.com/blog/gpudirect-storage/), KvikIO enables the GPU to read data directly into GPU memory, bypassing CPU memory.
Here is a minimal example of how to do this:

```python
import kvikio.zarr

with zarr.config.enable_gpu():
    store = kvikio.zarr.GDSStore(root="/tmp/air-temp.zarr")
    ds = xr.open_dataset(filename_or_obj=store, engine="zarr")
    assert isinstance(ds.air.data, cp.ndarray)
```

This will read the data directly from the Zarr store to GPU memory, significantly reducing I/O latency, especially for large datasets.
However, it relies on the [NVIDIA GPUDirect Storage (GDS)](https://docs.nvidia.com/datacenter/pgp/gds/index.html) feature to be enabled and correctly configured on your system.

> **Note**: Even with GDS, the decompression step is still occurs on the CPU (see next section for GPU solutions!). This means that the data is still being decompressed on the CPU before being transferred to the GPU. However, this is still a significant improvement over the previous method, as it reduces the amount of data that needs to be transferred over the PCIe bus. In the figure below, we show the flowchart of the data loading process with GDS enabled (i.e. using `kvikio`):

Reviewer 3: ```suggestion
**Note**: Even with GDS, the decompression step will still occur on the CPU (see next section for GPU solutions!). This means that the data is still being decompressed on the CPU before being transferred to the GPU. However, this is still a significant improvement over the previous method, as it reduces the amount of data that needs to be transferred over the PCIe bus. In the figure below, we show the flowchart of the data loading process with GDS enabled (i.e. using `kvikio`):
```

![Flowchart-technically decompression is still done on CPUs](/posts/gpu-pipline/flowchart_2.png)

### Step 3: GPU-based decompression with nvCOMP 🚀

For a fully GPU-native pipline, the decompression step should also be done on the GPU. This is where [NVIDIA's nvCOMP](https://developer.nvidia.com/nvcomp) library comes in. nvCOMP provides fast, GPU-native implementations of popular compression algorithms like Zstandard (Zstd)

> With nvCOMP, all steps of data loading including reading, decompressing, and transforming data can be done on the GPU, significantly reducing the time spent on data loading. Here is a flowchart of the data loading process with GDS and GPU-based decompression enabled:

Reviewer 3: ```suggestion
With nvCOMP, all steps of data loading including reading from disk, decompression, and transforming data can be done on the GPU, significantly reducing the time spent on data loading. Here is a flowchart of the data loading process with GDS and GPU-based decompression enabled:
```

![GPU native decompression](/posts/gpu-pipline/flowchart_3.png)

**Sending compressed instead of uncompressed data to the GPU means less data transfer overall, reducing I/O latency from storage to device.**

To unlock this, we would need zarr-python to support GPU-based decompression codecs, with one for Zstandard (Zstd) currently being implemented in [this PR](https://github.com/zarr-developers/zarr-python/pull/2863).

We tested the performance of GPU-based decompression using nvCOMP with Zarr v3 and KvikIO, and compared it to CPU-based decompression using [this data reading benchmark here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/benchmark/era5_zarr_benchmark.py).

Here are the results:

![GPU native decompression](/posts/gpu-pipline/zstd_benchmark.png)

> These results show that GPU-based decompression can significantly reduce the time spent on data loading and cut I/O latency from storage to device (less data transfer over PCIe/NVLink). This is especially useful for large datasets, as it allows for faster data loading and processing.

> Keep an eye on this space, as we are working on integrating this into the Zarr ecosystem to enable GPU-based decompression for Zarr stores. This will allow for a fully GPU-native workflow, where all steps of data loading, including reading, decompressing, and transforming data, can be done on the GPU.

Reviewer 3: ```suggestion
Keep an eye on this space, as we are working on integrating this into the Zarr ecosystem to enable GPU-based decompression for Zarr stores. This will allow for a fully GPU-native workflow, where all steps of data loading, including reading, decompression, and transforming data, can be done on the GPU.
```

Profiling results from the DALI pipeline demonstrate effective overlap between CPU and GPU workloads, significantly reducing GPU idle time (blue) and increasing overall training throughput:

<img
  src='/posts/gpu-pipline/profiling_screenshot_dali.png'
  alt='baseline plot'
/>

## Going Forward 🔮

This work is still ongoing, and we are continuing to explore ways to optimize data loading and processing for large-scale geospatial AI/ML workflows. We started this work during a 3-day hackathon, and we are excited to continue this work in the future. During the hackathon, we were able to make significant progress in optimizing data loading and processing for large-scale geospatial AI/ML workflows.

We are continuing to explore the following areas:

- GPU Direct Storage (GDS) for optimal performance
- NVIDIA DALI

> - Support for sharded Zarr with GPU-friendly access patterns [already merged in Zarr v3]().

Reviewer 3: ```suggestion
- Support for sharded Zarr with GPU-friendly access patterns already [merged](https://github.com/zarr-developers/zarr-python/pull/2978) in Zarr v3.0.8.
```

For this hackathon, we developed a benchmark of training a U-NET (with ResNet encoder) model on the ERA-5 Dataset to predict next time steps. The training pipeline used a standard PyTorch DataLoader and supported both single-GPU and multi-GPU training via Distributed Data Parallel (DDP). The full repo is available [here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus).

### Initial Performance Bottlenecks

First, we used NVIDIA's [Nsight Systems](https://developer.nvidia.com/nsight-systems) to profile our code and identify performance bottlenecks.

The initial profiling results clearly showed that the data loading step was the main bottleneck in our pipeline, with minimal overlap between CPU and GPU compute steps, which meant that the GPU was often idle while waiting for the CPU to load data.

Here are some screenshots of the profiling results:

<div style={{ display: 'flex', gap: '1rem', justifyContent: 'center' }}>
  <img
    src='/posts/gpu-pipline/profiling_screenshot1.png'
    alt='Issue 1'
    style={{ width: '50%' }}
  />
  <img
    src='/posts/gpu-pipline/profiling_screenshot2.png'
    alt='Issue 2'
    style={{ width: '55%' }}
  />
</div>

> We further quantified this bottleneck by comparing data loading and training throughput, as shown in the figure below:

Reviewer 4: ```suggestion
We further quantified this bottleneck by comparing data loading and training throughput, as shown in the figure below (higher bars/more throughput is better):
```

<img
  src='/posts/gpu-pipline/baseline.png'
  alt='baseline plot'
  style={{ display: 'inline-block', width: '50%', maxWidth: '400px' }}
/>

In the plot above, the three bars represent:

> - Baseline: Baseline throughput of the end-to-end pipeline using real data.

Reviewer 4: ```suggestion
- Training (real data): Baseline throughput of the end-to-end pipeline using real data.
```

- No Training (i.e. data loading throughput): Throughput of the data loading without any training (to measure the time spent on data loading vs. training).
- Synthetic Data (i.e. Training throughput): Throughput of the data loading using synthetic data (to remove the data loading bottleneck).

The results show that the data loading step is the main bottleneck in our pipeline, with **much** lower throughput compared to the training step.

## Hackathon: Strategies Explored!

During the hackathon, we tested the following strategies to improve the data loading performance:

1. **Optimized Chunking & Compression**
   - We explored different chunking and compression strategies to optimize the data loading performance. We found that using Zarr v3 with optimized chunking and compression significantly improved the data loading performance.

> 2. **GPU-native data loading with Zarr v3 and KvikIO**

Reviewer 4: ```suggestion
2. **GPU-native data loading with Zarr Python 3 and KvikIO**
```

>    - Leveraging Zarr v3's support for reading data directly into GPU memory using CuPy arrays, we utilized [KvikIO](https://docs.rapids.ai/api/kvikio/stable/) to bypass CPU memory, enabling direct data transfer from storage to GPU.

Reviewer 4: ```suggestion
   - Leveraging Zarr Python 3's support for reading data directly into GPU memory using CuPy arrays, we utilized [KvikIO](https://docs.rapids.ai/api/kvikio/stable/) to bypass CPU memory, enabling direct data transfer from storage to GPU.
```

3. **Using `nvcomp` for decompression on GPUs**
   - We explored using [NVIDIA's nvCOMP library](https://developer.nvidia.com/nvcomp) for GPU-accelerated decompression of Zarr data. This allowed us to offload the decompression step to the GPU, reducing the time spent on data loading.
4. **NVIDIA DALI**: We explored integrating [NVIDIA's Data Loading Library (DALI)](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/index.html) into our pipeline to facilitate efficient data loading and preprocessing directly on the GPU. NVIDIA DALI provides highly optimized building blocks and an execution engine for data processing, accelerating deep learning applications.

### Step 1: Optimized chunking & Compression

The ERA-5 dataset we were using had a sub-optimal chunking scheme of `{'time': 10, 'channel': C, 'height': H, 'width': W}`, which meant that a minimum of 10 timesteps of data was being read even if we only needed 2 consecutive timesteps at a time.

> We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr v3.

Reviewer 4: ```suggestion
We decided to rechunk the data to align with our access pattern of 1-timestep at a time, while reformating to Zarr format 3.
```

The full script is available [here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/rechunk/era5_rechunking.ipynb), with the main code looking like so:

```python
import xarray as xr

ds: xr.Dataset = xr.open_mfdataset("ERA5.zarr")
# Rechunk the data
ds = ds.chunk({"time": 1, "level": 1, "latitude": 640, "longitude": 1280})
# Save to Zarr v3
ds.to_zarr("rechunked_ERA5.zarr", zarr_version=3)
```

For more optimal performance, consider:

1. Storing the data without compression (if not transferring over a network), as decompressing data can slow down read speeds. But see also GPU decompression with nvCOMP below. 😉
2. Concatenating several data variables together **if** a single chunk size is too small (`<1MB`), at the expense of reducing readability of the Zarr store.
   Having too many small chunks can be detrimental to read speeds. A compressed chunk should be `>1MB`, `<100MB` for optimal reads.
   - Alternatively, wait for [sharding](https://zarr.readthedocs.io/en/stable/user-guide/performance.html#sharding) to be supported for GPU buffers in Zarr-python.
3. Align chunks with model access pattern.

> The plot below shows the performance of the original dataset vs. the rechunked dataset (to optimal chunk size) vs. uncompressed Zarr v3 dataset.

Reviewer 4: ```suggestion
The plot below shows the performance of the original dataset vs. the rechunked dataset (to optimal chunk size) vs. uncompressed Zarr format 3 dataset.
```

### Step 2: Reading with Zarr-Python v3 + kvikIO 📖

> [Zarr v3](https://zarr.dev/blog/zarr-python-3-release/) introduces new capabilities, including the ability to [read from Zarr stores into CuPy arrays (i.e. GPU memory)](https://github.com/zarr-developers/zarr-python/issues/2574).

Reviewer 4: ```suggestion
[Zarr Python 3](https://zarr.dev/blog/zarr-python-3-release/) introduces new capabilities, including the ability to [read from Zarr stores into CuPy arrays (i.e. GPU memory)](https://github.com/zarr-developers/zarr-python/issues/2574).
```

Specifically, you can use the [`zarr-python`](https://github.com/zarr-developers/zarr-python) driver to read data from zarr->CPU->GPU, or the [`kvikio`](https://github.com/rapidsai/kvikio) driver to read data from zarr->GPU directly!

To benefit from these new features, we recommend installing:

> - [`zarr>=3.0.3`](https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.3)

Reviewer 4: ```suggestion
- [`zarr>=3.0.8`](https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.8)
```

Author: Perfect. Thanks @weiji14 & @maxrjones

- [`xarray>=2025.03.00`](https://github.com/pydata/xarray/releases/tag/v2025.03.0)
- [`kvikio>=25.04.00`](https://github.com/rapidsai/kvikio/releases/tag/v25.04.00)

Reading to GPU can be enabled by using the [`zarr.config.enable_gpu()`](https://zarr.readthedocs.io/en/v3.0.6/user-guide/gpu.html) setting like so:

```python
import cupy as cp
import xarray as xr
import zarr

airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")
airt.to_zarr(store="/tmp/air-temp.zarr", mode="w", zarr_format=3, consolidated=False)

with zarr.config.enable_gpu():

>     ds = xr.open_dataset("/tmp/air-temp.zarr", engine="zarr", consolidated=False)

Reviewer 4: Can you explain why not to use consolidated metadata for direct to GPU reading? I am surprised about this because I would expect it to improve performance.

    assert isinstance(ds.air.data, cp.ndarray)
```

Note that using `engine="zarr"` like above would still result in data being loaded into CPU memory before being transferred to GPU memory.

> If your system supports [GPU Direct Storage (GDS)](https://developer.nvidia.com/blog/gpudirect-storage/), you can use the `kvikio` to read data directly into GPU memory, bypassing CPU memory.

Reviewer 4: ```suggestion
If your system supports [GPU Direct Storage (GDS)](https://developer.nvidia.com/blog/gpudirect-storage/), you can use `kvikio` to read data directly into GPU memory, bypassing CPU memory.
```

```python
import kvikio.zarr

with zarr.config.enable_gpu():
    store = kvikio.zarr.GDSStore(root="/tmp/air-temp.zarr")
    ds = xr.open_dataset(filename_or_obj=store, engine="zarr")
    assert isinstance(ds.air.data, cp.ndarray)
```

This will read the data directly from the Zarr store to GPU memory, significantly reducing I/O latency. This is especially useful for large datasets, as it reduces the amount of data that needs to be transferred between CPU and GPU memory, but requires GPU Direct Storage (GDS) to be enabled on your system.

While decompression still occurs on the CPU, this marks an important step toward enabling fully GPU-native workflows. In the figure below, we show the flowchart of the data loading process with GDS enabled.

![Flowchart-technically decompression is still done on CPUs](/posts/gpu-pipline/flowchart_2.png)

> Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:

Reviewer 4: ```suggestion
Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at [https://xarray.dev/blog/xarray-kvikio](https://xarray.dev/blog/xarray-kvikio)), this can be simplified to:
```

For a fully GPU-native workflow, we can let the GPU do all of the work, including decompression. This is where [NVIDIA's nvCOMP](https://developer.nvidia.com/nvcomp) library comes in. nvCOMP provides GPU-accelerated compression and decompression algorithms for various data formats, including Zstandard (Zstd). This means that all steps of data loading including reading, decompressing, and transforming data can be done on the GPU, significantly reducing the time spent on data loading.

![GPU native decompression](/posts/gpu-pipline/flowchart_3.png)

Sending compressed instead of uncompressed data to the GPU means less data transfer overall, reducing I/O latency from storage to device.
To unlock this, we would need zarr-python to support GPU-based decompression codecs, with one for Zstandard (Zstd) currently being implemented in [this PR](https://github.com/zarr-developers/zarr-python/pull/2863).

Figure above shows benchmark comparing CPU vs GPU-based decompression, with or without GDS enabled using [the data reading benchmark here](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/blob/main/benchmark/era5_zarr_benchmark.py).

> ![GPU native decompression](/posts/gpu-pipline/zstd_benchmark.png)

Reviewer 4: do you know why it's slower with GDS?

Author: Potentially GDS compat mode. I ran a new set of tests and updated the figure per our previous meeting.

        self.i = 0
        return self

    def __next__(self):
        batch, labels = [], []
        for _ in range(self.batch_size):
            idx = self.indices[self.i % len(self.images)]
            batch.append(self.images[idx])
            labels.append(self.labels[idx])
            self.i += 1
        return batch, labels
```

II. Define a DALI pipeline with `ExternalSource` operator to read data from the iterator.

```
eii = ExternalInputIterator()
pipe = dali.pipeline.Pipeline(batch_size=16, num_threads=4, device_id=0)

with pipe:
  images, labels = fn.external_source(
      source=eii,
      num_outputs=2,
      device="gpu", # use GPU memory
      batch_size=16,
  )
```

III. Build and run the pipeline:

```
pipe.build()
output = pipe.run()
images_gpu, labels_gpu = output
```

Next, checkout the [end-to-end example](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_ML_optimization) directory, where we show how to integrate the DALI pipeline into a PyTorch DataLoader and training loop. This example demonstrates how to use DALI to load data from Zarr stores, preprocess it on the GPU, and feed it into a PyTorch model for training.

> Profiling results from the DALI pipeline demonstrate effective overlap between CPU and GPU workloads, significantly reducing GPU idle time (blue) and increasing overall training throughput:

Reviewer 4: the GPU activity is blue, right? rather than the idle time

<img
  src='/posts/gpu-pipline/profiling_screenshot_dali.png'
  alt='baseline plot'
  style={{ display: 'inline-block', width: '70%', maxWidth: '400px' }}
/>

## Going Forward 🔮

This work is still ongoing, and we are continuing to explore ways to optimize data loading and processing for large-scale geospatial AI/ML workflows. We started this work during a 3-day hackathon, and we are excited to continue this work in the future. During the hackathon, we were able to make significant progress in optimizing data loading and processing for large-scale geospatial AI/ML workflows.

We are continuing to explore the following areas:

- GPU Direct Storage (GDS) for optimal performance
- NVIDIA DALI
- Support for sharded Zarr with GPU-friendly access patterns
- Work out how to use GDS when reading from cloud object store instead of on-prem disk
- GPU-based decompression with nvCOMP

> ## Lessons Learned 💡

> - **Chunking matters!** It really does and can make a huge difference in performance.

> > - **Zarr v3 enables GPU-native workflows**: Zarr v3 introduces experimental support for reading data directly into GPU memory via `zarr.config.enable_gpu()`. However, this is currently limited to the final stage of the codec pipeline, with decompression still handled by the CPU. We are working on enabling GPU-native decompression using `nvComp` to eliminate the host-device transfer.

Reviewer 4: ```suggestion
> - **Zarr Python 3 enables GPU-native workflows**: Zarr Python 3 introduces experimental support for reading data directly into GPU memory via `zarr.config.enable_gpu()`. However, this is currently limited to the final stage of the codec pipeline, with decompression still handled by the CPU. We are working on enabling GPU-native decompression using `nvComp` to eliminate the host-device transfer.
```

> - **GPU Direct Storage (GDS)** can be an improvement for data-intensive workflows, but requires some setup and configuration.

> > - **Compression trade-offs**: Using compression can reduce the amount of data transferred, but can also increase the time spent on decompression. We found that using Zarr v3 with GPU-based decompression can significantly improve performance.
> > - **NVIDIA DALI** is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.
> > - **CuPy-Xarray integration** is still a work in progress, but can be very useful for GPU-native workflows. Please see this PR for more details: [xarray-contrib/cupy-xarray#70](https://github.com/xarray-contrib/cupy-xarray/pull/70).
> > - **GPU-native decompression** is a promising area for future work, but full support (e.g. GPU-side Zstd decompression) requires further development and testing.

Reviewer 4: ```suggestion
> - **Compression trade-offs**: Using compression can reduce the amount of data transferred, but can also increase the time spent on decompression. We found that using Zarr v3 with GPU-based decompression can significantly improve performance.
> - **GPU-native decompression** is a promising area for future work, but full support (e.g. GPU-side Zstd decompression) requires further development and testing.
> - **NVIDIA DALI** is a powerful tool for optimizing data loading, but requires some effort to integrate into existing workflows.
> - **CuPy-Xarray integration** is still a work in progress, but can be very useful for GPU-native workflows. Please see this PR for more details: [xarray-contrib/cupy-xarray#70](https://github.com/xarray-contrib/cupy-xarray/pull/70).
> - **NVIDIA Nsight** provides a [powerful tool](https://developer.nvidia.com/nsight-systems) for identifying bottlenecks.
```
